File: 프로그래머스/1/172928. 공원 산책/공원 산책.java
class Solution {
    public int[] solution(String[] park, String[] routes) {
        int scount[] = new int[2];
        boolean check[][] = new boolean[park.length][park[0].length()];
        for(int i=0; i<park.length; i++){
            int len = park[i].length();
            for(int j=0; j<len; j++){
                if(park[i].charAt(j) == 'S'){
                    scount[0] = i;
                    scount[1] = j;
                }
                if(park[i].charAt(j) == 'X'){
                    check[i][j] = true;
                }
            }
        }
        for (int i = 0; i < routes.length; i++) {
            String tmp = routes[i];
            String date[] = tmp.split(" ");
            int tmp2 = Integer.parseInt(date[1]);
            int newX = scount[0];
            int newY = scount[1];

            if (date[0].equals("N")) {
                newX -= tmp2;
            } else if (date[0].equals("S")) {
                newX += tmp2;
            } else if (date[0].equals("W")) {
                newY -= tmp2;
            } else {
                newY += tmp2;
            }
            if (newX >= 0 && newX < park.length && newY >= 0 && newY < park[newX].length() && !hasXInPath(scount[0], scount[1], newX, newY, park, check)) {
                scount[0] = newX;
                scount[1] = newY;
            }
        }
        return scount;
    }
     private boolean hasXInPath(int startX, int startY, int endX, int endY, String[] park, boolean[][] check) {
        int stepX = (startX < endX) ? 1 : (startX > endX) ? -1 : 0;
        int stepY = (startY < endY) ? 1 : (startY > endY) ? -1 : 0;
        int currentX = startX;
        int currentY = startY;

        while (currentX != endX || currentY != endY) {
            currentX += stepX;
            currentY += stepY;
            if (check[currentX][currentY]) {
                return true;
            }
        }

        return false;
    }
}

File: 프로그래머스/1/172928. 공원 산책/공원 산책.cpp
#include <string>
#include <vector>
#include<iostream>
using namespace std;

vector<int> solution(vector<string> park, vector<string> routes) {
    vector<int> answer(2,0);
    int xlen = park.size();
    int ylen = park[0].length();
    for(int i=0; i<park.size(); i++){
        for(int j=0; j<park[i].length(); j++){
            if(park[i][j] == 'S'){
                answer[0] = i;
                answer[1] = j;
            }
        }
    }
    for(int j=0; j<routes.size(); j++){
        char fir = routes[j][0];
        int sec = routes[j][2] - '0';
        bool check = true;
        
        switch(fir){
            case 'N' :
                if(answer[0] - sec >= 0){
                    for(int i=1; i<=sec; i++){
                        if(park[answer[0] - i][answer[1]] == 'X'){
                            check = false;
                            break;
                        }
                    }
                }
                else{
                    check = false;
                }
                if(check){
                    answer[0] -= sec;
                }
                break;
            case 'S' :
                if(answer[0] + sec < xlen){
                    for(int i=1; i<=sec; i++){
                        if(park[answer[0] + i][answer[1]] == 'X'){
                            check = false;
                            break;
                        }
                    }
                }
                else{
                    check = false;
                }
                if(check){
                    answer[0] += sec;
                }
                break;
            case 'W' :
                if(answer[1] - sec >= 0){
                    for(int i=1; i<=sec; i++){
                        if(park[answer[0]][answer[1] - i] == 'X'){
                            check = false;
                            break;
                        }
                    }
                }
                else{
                    check = false;
                }
                if(check){
                    answer[1] -= sec;
                }
                break;
            case 'E' :
                if(answer[1] + sec < ylen){
                    for(int i=1; i<=sec; i++){
                        if(park[answer[0]][answer[1] + i] == 'X'){
                            check = false;
                            break;
                        }
                    }
                }
                else{
                    check = false;
                }
                if(check){
                    answer[1] += sec;
                }
                break;
        }
    }
    return answer;
}